### PR TITLE
Implement quaternion

### LINKED ISSUE #17: Implement Quaternion
A list of features to implement
- [ ] Get axis
- [ ] Get angle
- [ ] Get axis/angle
- [ ] Normalized
- [ ] Normalize
- [ ] Set identity
- [ ] Inverse
- [ ] Binary operators
  - [ ] Multiply

### PR DESCRIPTION
**Pull request by hyp3rflow**
This revision includes:
- Implement `Quaternion` and related test code (#17)
  - get_vector_part
  - get_scalar_part
  - set_identity
  - make_identity
  - cale
  - dot
  - inverse
  - conjugate
  - get_squared_length
  - get_length
  - rotate
  - from_euler_angles
  - from_axis_angle
  - get_axis_angle
  - add
  - mul

### CODE REVIEW

File: oxide_core/quaternion.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+pub struct Quaternion {
+  pub x: f32,
+  pub y: f32,
+  pub z: f32,
+  pub w: f32,
+}
+
+pub trait VectorPart {
+  fn get_vector_part(&self) -> Vector3;
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: Is there a reason why VectorPart should be a trait?

File: oxide_core/quaternion.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      x: self.x,
+      y: self.y,
+      z: self.z,
+    }
+  }
+}
+
+pub trait ScalarPart {
+  fn get_scalar_part(&self) -> f32;
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: Same as above.

File: oxide_core/quaternion.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  let one = 1.0f32;
+  let zero = 0.0f32;
+
+  Quaternion {
+    x: zero,
+    y: zero,
+    z: zero,
+    w: one,
+  }
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: What does the function name `id()` mean?

File: oxide_core/quaternion.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+impl Quaternion {
+  pub fn get_squared_length(&self) -> f32 {
+    self.w * self.w + self.get_vector_part().get_squared_length()
+  }
+  pub fn get_length(&self) -> f32 {
+    self.get_squared_length().sqrt()
+  }
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: I think that these functions don't need.

File: oxide_core/quaternion.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+pub fn rotate_vector(q: Quaternion, v: Vector3) -> Vector3 {
+  let vector_q = q.get_vector_part();
+  let t = vector_q.outer_product(&v).scale(2.0f32);
+  v + t.scale(q.w) + vector_q.outer_product(&t)
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: Can you replace this function with `rotate(angle_in_radius: f32)` like this?
```C++
template <typename T>
void Quaternion<T>::Rotate(T angleInRadians)
{
    Vector3<T> axis;
    T currentAngle;

    GetAxisAngle(&axis, &currentAngle);

    currentAngle += angleInRadians;

    Set(axis, currentAngle);
}
```

**Review by utilForever — CHANGES_REQUESTED** [review]
You should check my review. 🍡

File: oxide_core/quaternion.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
+use crate::vector::vector3::*;
+use crate::vector::*;
+use std::ops;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
```
Comment (BeLeap) [inline oxide_core/quaternion.rs]: Wouldn't it be better to add Serde?

**Review by BeLeap — CHANGES_REQUESTED** [review]

File: oxide_core/quaternion.rs (hunk 1 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: Oh, I'll remove it and use just ```impl``` without making traits.

File: oxide_core/quaternion.rs (hunk 3 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: It means identity quaternion. Should I have to rename it for making it clear?

File: oxide_core/quaternion.rs (hunk 4 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: I think so. I'll remove it.

File: oxide_core/quaternion.rs (hunk 5 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: Is it okay to change like this?
```suggestion
pub fn rotate(q: Quaternion, angle_in_radius: f32) -> Quaternion {
  let (axis, angle) = get_axis_angle(q);
  from_axis_angle(axis, angle + angle_in_radius)
}
```

File: oxide_core/quaternion.rs (hunk 6 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: Okay. I'll add it.

File: oxide_core/quaternion.rs (hunk 3 above)
Comment (utilForever) [reply oxide_core/quaternion.rs]: I think that this function should be removed.
Instead, I suggest the function `set_identity()` or `make_identity()` like this:
```C++
template <typename T>
void Quaternion<T>::SetIdentity()
{
    Set(1, 0, 0, 0);
}
```
or
```C++
template <typename T>
Quaternion<T> Quaternion<T>::MakeIdentity()
{
    return Quaternion{};
}
```

File: oxide_core/quaternion.rs (hunk 5 above)
Comment (utilForever) [reply oxide_core/quaternion.rs]: Yes, good.

File: oxide_core/quaternion.rs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  let half_angle = angle / 2.0f32;
+  let vector_part = axis.scale(half_angle.sin());
+
+  Quaternion {
+    x: vector_part.x,
+    y: vector_part.y,
+    z: vector_part.z,
+    w: half_angle.cos(),
+  }
+}
```
Comment (utilForever) [inline oxide_core/quaternion.rs]: I suggest `from_axis_angle(axis: Vector3, angle: f32) -> Quaternion` → `set_axis_angle(axis: Vetor3, angle: f32) -> ()`. I think that this function don't need return type.

**Review by utilForever — CHANGES_REQUESTED** [review]
I reviewed another one. Please check it. 🧷

File: oxide_core/quaternion.rs (hunk 3 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: I'll implement both.

File: oxide_core/quaternion.rs (hunk 1 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

File: oxide_core/quaternion.rs (hunk 2 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

File: oxide_core/quaternion.rs (hunk 3 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)
Test implemented., [`b2c3eee` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/b2c3eee61212a1ec5c14f8b1960c285eb70639e8)

File: oxide_core/quaternion.rs (hunk 4 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: I noticed that function ```get_length``` is needed to implement ```inverse``` function.
So, I changed the visibility of those two functions.
✅ [`bce2141` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/bce214119efc72065ee66d8ddc01cdc8c11c529f)

File: oxide_core/quaternion.rs (hunk 5 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

File: oxide_core/quaternion.rs (hunk 6 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

File: oxide_core/Cargo.toml [hunk 8]
```diff
@@ -10,7 +10,7 @@ license = "MIT"
 edition = "2018"
 
 [lib]
-name = "oxide_render"
+name = "oxide_core"
```
Comment (BeLeap) [inline oxide_core/Cargo.toml]: I think we should fix this.

**Review by BeLeap — CHANGES_REQUESTED** [review]

File: oxide_core/Cargo.toml (hunk 8 above)
Comment (hyp3rflow) [reply oxide_core/Cargo.toml]: Isn't it already fixed?

Comment (BeLeap) [reply oxide_core/Cargo.toml]: It meant moving from oxide_core to oxide_math.

File: oxide_math/quaternion.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        w: half_angle.cos(),
+    }
+}
+
+pub fn get_axis_angle(q: Quaternion) -> (Vector3, f32) {
+    let axis = q.get_vector_part().normalize();
+    let angle = 2.0 * q.get_scalar_part().acos();
+
+    if angle > ::std::f32::consts::PI {
+        (axis.negate(), 2.0 * ::std::f32::consts::PI - angle)
```
Comment (BeLeap) [inline oxide_math/quaternion.rs:131]: Add test for this line

**Review by BeLeap — CHANGES_REQUESTED** [review]

File: oxide_core/quaternion.rs (hunk 7 above)
Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: ㅎㅎ..

Comment (hyp3rflow) [reply oxide_core/quaternion.rs]: 일단 axis랑 angle을 이용해서 quaternion 값을 만든다는 것이 동일한 것 같은데, 그 작업 자체는 기존에 어떤 값이 있는지에 대해 영향을 받지 않기 때문에 ```from_axis_angle```로 만들었습니다. @utilForever

Comment (utilForever) [reply oxide_core/quaternion.rs]: Okay. I understand.

**Review by utilForever — CHANGES_REQUESTED** [review]
Please to add description (See #13)

File: oxide_math/quaternion.rs (hunk 9 above)
Comment (hyp3rflow) [reply oxide_math/quaternion.rs:131]: ✅ [`8923709` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/8923709bec2cecf561cddb7441a4fbd3e0a9605a)

**Comment by hyp3rflow** [comment]
I filled in the description. @utilForever

**Review by utilForever — APPROVED** [review]
LGTM now. Thanks for your work! 🕊️
Note that I update CI steps to improve test coverage. 🦆

**Comment by BeLeap** [comment]
LGTM

**Review by BeLeap — APPROVED** [review]
LGTM🛫

**Comment by LuneFleurs** [comment]
LGTM

**Review by LuneFleurs — APPROVED** [review]
LGTM Good change!

**Pull request merged**